From 14a18fa555a9461c204955c62c8d7e6330984eaf Mon Sep 17 00:00:00 2001
From: Elizabeth Figura <zfigura@codeweavers.com>
Date: Fri, 17 Oct 2025 17:32:00 -0500
Subject: [PATCH 25/48] ntdll: Handle reparse points in NtQueryDirectoryFile().

(cherry picked from commit 513ada7c989bf5be4e55273f4c320dd392cd12cb)

CW-Bug-Id: #26492
---
 dlls/ntdll/tests/file.c | 13 +++++--------
 dlls/ntdll/unix/file.c  |  1 +
 2 files changed, 6 insertions(+), 8 deletions(-)

diff --git a/dlls/ntdll/tests/file.c b/dlls/ntdll/tests/file.c
index 1b25d7b2fec..ed368a10b23 100644
--- a/dlls/ntdll/tests/file.c
+++ b/dlls/ntdll/tests/file.c
@@ -6448,14 +6448,11 @@ static void test_reparse_points(void)
 
     swprintf( path, ARRAY_SIZE(path), L"%s/testreparse_dirlink", temp_path );
     find_handle = FindFirstFileW( path, &find_data );
-    todo_wine ok( find_handle != INVALID_HANDLE_VALUE, "got error %lu\n", GetLastError() );
-    if (find_handle != INVALID_HANDLE_VALUE)
-    {
-        todo_wine ok( find_data.dwFileAttributes == (FILE_ATTRIBUTE_DIRECTORY | FILE_ATTRIBUTE_REPARSE_POINT),
-            "got attributes %#lx\n", find_data.dwFileAttributes );
-        todo_wine ok( find_data.dwReserved0 == IO_REPARSE_TAG_MOUNT_POINT, "got tag %#lx\n", find_data.dwReserved0 );
-        FindClose( find_handle );
-    }
+    ok( find_handle != INVALID_HANDLE_VALUE, "got error %lu\n", GetLastError() );
+    todo_wine ok( find_data.dwFileAttributes == (FILE_ATTRIBUTE_DIRECTORY | FILE_ATTRIBUTE_REPARSE_POINT),
+        "got attributes %#lx\n", find_data.dwFileAttributes );
+    todo_wine ok( find_data.dwReserved0 == IO_REPARSE_TAG_MOUNT_POINT, "got tag %#lx\n", find_data.dwReserved0 );
+    FindClose( find_handle );
 
     /* Test using the reparse point as a parent.
      * On some machines this returns STATUS_REPARSE_POINT_NOT_RESOLVED (which
diff --git a/dlls/ntdll/unix/file.c b/dlls/ntdll/unix/file.c
index 995fc2002a4..2a2620e4c21 100644
--- a/dlls/ntdll/unix/file.c
+++ b/dlls/ntdll/unix/file.c
@@ -1603,6 +1603,7 @@ static BOOL append_entry( struct dir_data *data, const char *long_name,
 
     long_len = ntdll_umbstowcs( long_name, strlen(long_name), long_nameW, ARRAY_SIZE(long_nameW) );
     if (long_len == ARRAY_SIZE(long_nameW)) return TRUE;
+    if (long_nameW[long_len - 1] == '?') --long_len;
     long_nameW[long_len] = 0;
 
     if (short_name)
-- 
2.53.0

